### PR TITLE
Attempt to detect --distro automatically

### PR DESCRIPTION
**Pull request by praiskup**
Fixes: #64

### CODE REVIEW

**Comment by praiskup** [comment]
I know this is not exactly what we talked about in #64, but since the default='auto'
was sort of OK according to @bkabrda POV, it doesn't seem to be too different...

Simply, if --distro is set explicitly, it is used, otherwise attempt to autodetection is done,
with fallback to failure saying that `--distro` is mandatory option.

**Comment by praiskup** [comment]
And, this is WIP, if it is OK to be pushed -- I'll add add comprehensive unittests as possible.

**Comment by bkabrda** [comment]
LGTM :+1:

**Comment by praiskup** [comment]
Do you want to have another look?

File: tests/unittests/test_distro_detection.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+from distgen.distro_version import detect_default_distro
+
+from mock import patch
+
+class TestDistroDetection(object):
+
+    @patch('subprocess.Popen')
+    @patch('platform.dist')
+    def test_rpm(self, dist, popen):
+        for arch in [b'x86_64', b'i386']:
```
Comment (bkabrda) [inline tests/unittests/test_distro_detection.py]: For testing multiple similar scenarios, I'd suggest using `pytest.mark.parametrize`. It makes tests cleaner and executes them as separate test cases. See e.g. https://github.com/devexp-db/distgen/blob/master/tests/unittests/test_commands.py#L18

Comment (praiskup) [reply tests/unittests/test_distro_detection.py]: yay, thanks for the point!

**Comment by praiskup** [comment]
ptal

File: distgen/distro_version.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        if name.lower() == 'rawhide':
+            version = 'rawhide'
+    elif os in ['centos', 'rhel']:
+        pass
+    else:
+        return None
+
+    # Only RPM distros for now:
+    cmd = ['rpm', '--eval', '%_arch']
+    sp = subprocess.Popen(cmd, stdout=subprocess.PIPE)
```
Comment (TomasTomecek) [inline distgen/distro_version.py]: there is a possible race here: you should explicitly wait for the command to finish, I encourage to use wrapper `subprocess.check_output` for this

**Comment by bkabrda** [comment]
LGTM, but @TomasTomecek is right about the potential race condition. Please fix that before we can merge.

File: distgen/distro_version.py (hunk 2 above)
Comment (praiskup) [reply distgen/distro_version.py]: Can you elaborate on the race?  I'm just curious..

Nvmd though, check_output is really even semantically better.

**Comment by praiskup** [comment]
s/Popen/check_output/ done, thanks for the review @TomasTomecek and @bkabrda !

File: distgen/distro_version.py (hunk 2 above)
Comment (bkabrda) [reply distgen/distro_version.py]: If the subprocess didn't finish before you tried to read its output, you'd read empty string (or partial result). `check_output` explicitly waits for the subprocess to finish and then returns whole output.

Comment (praiskup) [reply distgen/distro_version.py]: But the docs [1] claim that read() without argument should read the entire file.
[1] https://docs.python.org/3/tutorial/inputoutput.html#methods-of-file-objects

Comment (bkabrda) [reply distgen/distro_version.py]: Ok, I can't access the old version of the code any more and was assuming you were trying to do something else. I'm not sure then, maybe Tomas will explain.

Comment (praiskup) [reply distgen/distro_version.py]: It's actually instance of https://docs.python.org/3/library/io.html#io.BufferedReader

**Comment by TomasTomecek** [comment]
Oh, we talked about this offline and Pavel convinced me he had it right. I'm actually puzzled whether I was right it's a race.

**Pull request merged**